### PR TITLE
Avoid ordering peers based on peerid in blockfetch

### PR DESCRIPTION
**Pull request by karknu**
Peers where ordered based on SockAddr, which generally meant that all
peers would pick the same second choice peer to download blocks from in
Deadline mode. And the same third pick and so on.

This changes ranking so that peers are sorted based on G.

### CODE REVIEW

File: ouroboros-network/src/Ouroboros/Network/BlockFetch/DeltaQ.hs [hunk 1]
```diff
@@ -69,6 +70,27 @@ comparePeerGSV activePeers salt (a, a_p) (b, b_p) =
...
+      ( Hashable peer
+      )
+      => Int
+      -> (PeerGSV, peer)
+      -> (PeerGSV, peer)
+      -> Ordering
+comparePeerGSV' salt (a, a_p) (b, b_p) =
+    let gs_a = gs a
+        gs_b = gs b in
+    if abs (gs_a - gs_b) < 0.05*gs_a
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/BlockFetch/DeltaQ.hs:85]: What about `abc (gs_a - gs_b) < 0.05 * (max gs_a gs_b)` (or `min gs_a gs_b`).  If this condition is not symmetric `compareGSV'` might not satisfy `Ord` axioms.

Could you add a test which verifies `Ord` [axioms](https://hackage.haskell.org/package/base-4.16.0.0/docs/Prelude.html#t:Ord).

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/Types.hs [hunk 2]
```diff
@@ -613,6 +613,8 @@ data TracePeerSelection peeraddr =
      | TraceGovernorWakeup
      | TraceChurnWait          DiffTime
      | TraceChurnMode          ChurnMode
+     | TracePeerNess           Int (Double, Double, Double, Double, Double)
+                               Int (Double, Double, Double, Double, Double)
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/Types.hs:617]: Please add a haddock which explains the fields.

**Review by coot — CHANGES_REQUESTED** [review]

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/Types.hs (hunk 2 above)
Comment (karknu) [reply ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/Types.hs:617]: I will remove the commit. It doesn't belong in this PR.

File: ouroboros-network/src/Ouroboros/Network/BlockFetch/DeltaQ.hs (hunk 1 above)
Comment (karknu) [reply ouroboros-network/src/Ouroboros/Network/BlockFetch/DeltaQ.hs:85]: You're right. This bug also affects `compareGSV`. I'll implement your fix above and add some tests.

File: ouroboros-network/test/Test/Ouroboros/Network/BlockFetch.hs [hunk 3]
```diff
@@ -663,6 +668,46 @@ prop_terminate (TestChainFork _commonChain forkChain _forkChain) (Positive (Smal
...
+      unPeerGSVT :: PeerGSV
+    } deriving Show
+
+instance Arbitrary PeerGSVT where
+    arbitrary = do
+        gIn <- arbitrary
+        gOut <- arbitrary
+        let gsvIn  = ballisticGSV (unNetworkDelay gIn)  2e-6 (degenerateDistribution 0)
+            gsvOut = ballisticGSV (unNetworkDelay gOut) 2e-6 (degenerateDistribution 0)
+        return $ PeerGSVT $ PeerGSV (Time 0) gsvOut gsvIn
```
Comment (coot) [inline ouroboros-network/test/Test/Ouroboros/Network/BlockFetch.hs:690]: Could you add a shrinker?

File: ouroboros-network/test/Test/Ouroboros/Network/BlockFetch.hs [hunk 4]
```diff
@@ -663,6 +668,46 @@ prop_terminate (TestChainFork _commonChain forkChain _forkChain) (Positive (Smal
...
+         LT -> comparePeerGSV peerSet salt (b, pb) (a, pa) === GT
+         GT -> comparePeerGSV peerSet salt (b, pb) (a, pa) === LT
+         EQ -> comparePeerGSV peerSet salt (b, pb) (a, pa) === EQ
+
+-- | Check that identical peers are equal
+prop_comparePeerGSVEq :: Int -> Int -> PeerGSVT -> Bool -> Property
+prop_comparePeerGSVEq salt p (PeerGSVT a) aActive =
+    let peerSet = if aActive then Set.singleton p
+                             else Set.empty in
+    comparePeerGSV peerSet salt (a, p) (a, p) === EQ
```
Comment (coot) [inline ouroboros-network/test/Test/Ouroboros/Network/BlockFetch.hs:710]: Transitivity is missing, e.g. `a ≤ b ≤ c ⇒ a ≤ c`

File: ouroboros-network/test/Test/Ouroboros/Network/BlockFetch.hs [hunk 5]
```diff
@@ -663,6 +668,46 @@ prop_terminate (TestChainFork _commonChain forkChain _forkChain) (Positive (Smal
...
+        m <- choose (1, 1000 :: Int) -- A delay between 1 and 1000 ms
+        return $ NetworkDelay $ (fromIntegral m) / 1000
+
+newtype PeerGSVT = PeerGSVT {
+      unPeerGSVT :: PeerGSV
+    } deriving Show
+
+instance Arbitrary PeerGSVT where
+    arbitrary = do
+        gIn <- arbitrary
```
[the remark this answers is not in the document]
Comment (karknu) [reply ouroboros-network/test/Test/Ouroboros/Network/BlockFetch.hs]: Will switch to Delay.

File: ouroboros-network/test/Test/Ouroboros/Network/BlockFetch.hs (hunk 3 above)
Comment (karknu) [reply ouroboros-network/test/Test/Ouroboros/Network/BlockFetch.hs:681]: Sure.

File: ouroboros-network/test/Test/Ouroboros/Network/BlockFetch.hs (hunk 4 above)
Comment (karknu) [reply ouroboros-network/test/Test/Ouroboros/Network/BlockFetch.hs:706]: `comparePeerGSV` only supports transitivity when a << b << c.

**Review by coot — APPROVED** [review]
lgtm

**Comment by coot** [comment]
bors merge

**Comment by karknu** [comment]
bors cancel

**Comment by karknu** [comment]
Fixes #3536

**Pull request merged**